Terasology/WoodAndStone
# Added JumpSpeedPotion, and events for drinking potions.

A smaller update compared to the previous two. This was made possible by the hard work of @Josharias.

# Description of Changes

## Additions

**Potion Prefabs:**
* JumpSpeedPotion (doubles your jumping height for 10 seconds)

**Recipe Prefabs:**
* RecipeJumpSpeedPotion1 (1 WalkSpeedHerb and 1 SwimSpeedHerb)

**Jump Potion Related:**
* JumpSpeedAlterationEffect (to manage the jumping speed alteration effect)
* JumpSpeedComponent (to store the magnitude of the effect)

**Events:**
* BeforeDrinkPotionEvent (similar to how BeforeHealEvent works)
* DrinkPotionEvent

**Systems:**
* MobilityAlterationSystem (temporary until I move the event handler to the AlterationEffects module, along with JumpSpeedAlterationEffect and JumpSpeedComponent )

Image for JumpSpeedPotion1, which was created by me.

## Modifications

The MULTI_JUMP constant was added to PotionCommonEffects for future advancements.

The DrinkPotionAuthoritySystem was slightly overhauled to take advantage of the DrinkPotionEvent and BeforeDrinkPotionEvent. This should allow for other modules to influence the magnitude and duration of the potion's effects. The expireEffects method will also have to moved to the AlterationEffects module in the near future.

## Deletions

DebugHerb prefab removed.

# Known Isuses

* The burner for either the PortableCookingStation or PortableHerbalismStation may suddenly cease to work, ignoring any extra fuel added to the fuel slot.
* The getEntityBlocks error is caused by the method trying to look for either a BlockComponent or a BlockRegionComponent. Neither of these exist for the portable stations, as they are items.
* The first potionConsumed event handler method in the DrinkPotionAuthoritySystem will fail to work with the predefined potion types due to lack of a proper Genome. Need to add failure checks.

# How to Test

As this is the third set of revisions, there are several items that may not function as intended. Please report issues as you find them.

* Using the instructions given in the previous commit's "Herbalism" section, spawn several SwimSpeedHerbs and WalkSpeedHerbs.
* Make sure that each herb has a unique picture.
* Spawn and prepare the PortableHerbalismStation as usual (add water and fuel).
* Add all of the ingredients into the ingredients slots.
* Take a look at the ingredients list to ensure that three kinds of potions (JumpSpeed, SwimSpeed, and WalkSpeed) are listed.
* Begin crafting several JumpSpeed potions.
* If the burner suddenly breaks down (known issue), remove all of the items from the PortableHerbalismStation, drop it into the world, and repeat step 3.
* Once all the potions are crafted, begin drinking one of the JumpSpeed potions and jump up to verify if the effect matches the above description. You'll probably want to bring some HealPotions or use console commands.
* If all above checks work, the test was a success.

MORE TO BE ADDED.

# Credits

I would like to thank @Cervator, @Josharias, @SkySom, and @portokaliu for helping me throughout the development.
